rfcs/2022-04-20-12187-log-namespacing.md

```diff
@@ -0,0 +1,422 @@
+
+## Goals
+
+
+1. The schema / type should be known in all cases when using the "Vector" namespace
+2. The "Global Log Schema" should be ignored
+3. Users can opt in to the new namespace on a per-source basis (as well as globally)
+
+## Proposal
+
+### Changes
+
+Support will be added for different log namespaces in Vector. The default will be a "Legacy" namespace, which keeps
+behavior the same as it was before namespaces were introduced. A new "Vector" namespace will be added with the changes
+described below. The goal is to allow an opt-in migration to the new namespace. Eventually the legacy namespace
+can be deprecated and removed.
+
+The user-facing configuration for setting the log namespace will start with a simple `log_namespace` boolean setting.
+This will be available as a global setting and a per-source setting, both defaulting to false.
+A value of false means the "Legacy" namespace is used, true means the "Vector" namespace is used.
+This will seem like you are just enabling / disabling the "log namespace" feature. However, it leaves
+the option open in the future to allow string values and pick the name of the namespace if more namespaces are added.
+
+The "Global Log Schema" will be ignored when the Vector namespace is used. Instead of user-configurable keys, static
+keys will be used. Previously this was useful so users could choose names that didn't collide with other data in the event.
+With namespacing, this is no longer a concern.
+
+Similar to the Global Log Schema, some sources allow choosing key names where some data will be stored. This will also be
+removed when the "Vector" namespace is being used, in favor of static names (usually what the default was). An example
+is the "key_field" and "headers_key" in the kafka source.
+
+Many transforms / sinks rely on the "Global Log Schema" to get/modify info such as the timestamp. Since the log
+schema should be customizable per source, that means the transforms and sinks need to know which log namespace
+is being used by each log. In order to accomplish this, the `log_namespace` event metadata field will be set to
+`vector` if the new namespace is being used.
+
+
+
+### Types of Data
+
+#### Data
+
+This is the main content of a log.  It will be placed at the root of the event.
+What is considered the "main content" here is not very well-defined, but should generally be what most people would expect
+to receive from the source. It is not necessarily the "log message" itself.
+
+A good example is the "Datadog Agent" source. The root is the fields received by the Datadog agent. The log content is
+nested.
+
+Others such as the "socket" source will have the log content directly at the root.
+
+There is a special case when the codec is "bytes", since the data is just a string in that case. That means
+that the root is a string. Historically this has been forbidden, but it is possible to allow this.
+An example of this is the "socket" source, with the "bytes" codec.
```

Comment:
How does this interact with the Vector Metadata?

Reply:
They all go to the same place. I mention it a bit in the `Event Metadata` and `Additional nesting in event metadata` sections. But basically, we get to pick the name for all fields in there, so we can prevent collisions. But it might get a bit cozy.